File: server/internal/ports/job.go
package ports

import (
	"context"
	"time"

	"github.com/google/uuid"

	"github.com/EnsoTech-Studio/CryptoBot/server/internal/domain/job"
)

type JobDispatcher interface {
	Enqueue(context.Context, job.BacktestJob) error
	Claim(context.Context, string, time.Duration) (job.BacktestJob, error)
	Complete(context.Context, uuid.UUID) error
	Fail(context.Context, uuid.UUID, error, bool) error
}

File: app/ports/job.py
"""Job dispatcher seam. Mirrors `server/internal/ports/job.go`."""

from __future__ import annotations

from datetime import timedelta
from typing import Protocol
from uuid import UUID

from ..domain.job import BacktestJob


class JobDispatcher(Protocol):
    def enqueue(self, job: BacktestJob) -> None: ...

    def claim(self, worker_id: str, lease: timedelta) -> BacktestJob: ...

    def complete(self, job_id: UUID) -> None: ...

    def fail(self, job_id: UUID, err: Exception, retryable: bool) -> None: ...
